### DISCUSSION TITLE
Docker Hub Pull Rate Limits

### DISCUSSION THREAD

**Discussion by trademark18** [containrrr/watchtower#668] [Q&A] [answered]
Hi all!  My business uses watchtower to keep containers up to date on IoT devices.  We've been using it for years with good success, so thank you to the contributors out there!

Today I learned that Docker Hub is introducing a `pull` rate limit for Free/Anonymous users.  It's really low for anyone using watchtower to check for updates every few minutes.  This has prompted some questions I hope you can help me with:

1. Is watchtower's polling of the registry truly using a soon-to-be-rate-limited `docker pull`?  That may be obvious, but I just want to confirm.
2. If I do upgrade to a pro account to get more `pull`s, what changes do I need to make to watchtower to have it do those pulls in an authenticated way?
3. What other steps do I need to take to make sure I can keep using watchtower with Docker Hub?
4. I assume that the limit will apply per image, not per pulling IP.  If that's not right, let me know :)

Thanks in advance!

https://docs.docker.com/docker-hub/download-rate-limit/

**Comment by trademark18** [comment]
A colleague just sent me [this link from Docker Hub](https://www.docker.com/blog/understanding-inner-loop-development-and-pull-rates/) which states

> Docker defines pull rate limits as the number of manifest requests to Docker Hub. Rate limits for Docker pulls are based on the account type of the user requesting the image – not the account type of the image’s owner. For anonymous (unauthenticated) users, pull rates are limited based on the individual IP address. 

So it sounds like maybe we're in the clear as long as no watchtower is configured to update more frequently than every 5 minutes.  Can anyone sanity check this understanding and conclusion?

**Reply by simonaronsson** [reply]
Unfortunately not, as a pull interval of 5 minutes would generate `6*60/5` manifest requests per container.
For 5 containers that would mean 360 requests per 6 hour window.

However, if the rate limit applies on the manifest requests level, I assume skopeo won't help either? What are your conclusions from the material linked, @containrrr/watchtower-contributors?

**Reply by trademark18** [reply]
Hmm, so I suppose Docker Hub can only see the public IP, which would be shared among all devices on the network.  At one location we have 45 devices, so that one public IP must be issuing 540 requests per hour (3240 per 6 hour period).  Sooooo that's gonna be an issue.

**Reply by simonaronsson** [reply]
> Hmm, so I suppose Docker Hub can only see the public IP, which would be shared among all devices on the network. At one location we have 45 devices, so that one public IP must be issuing 540 requests per hour (3240 per 6 hour period). Sooooo that's gonna be an issue.

See below :)

**Comment by simonaronsson** [comment]
I've been digging through the docker daemon code for a bit now to get a clear understanding of how things work. From what I can tell, the only reason watchtower should ever be rate limited is if you actually get more than 100 updated images during a 6 hour window.

#### Let's break it down:

First, an HTTP HEAD request is done towards `/v2/<name>/manifests/<tag>/`, which returns a SHA digest if the tag exists. This request *does not* count towards the rate limit.

https://github.com/moby/moby/blob/bb23f1bf61cb49c5aa3f7df934d79c305afb7c8c/vendor/github.com/docker/distribution/registry/client/repository.go#L296-L348

The docker daemon then checks if the digest matches what we already have, if it does - it bails out without doing anything. 

https://github.com/moby/moby/blob/bb23f1bf61cb49c5aa3f7df934d79c305afb7c8c/distribution/pull_v2.go#L567-L571

If the digest differs from what we have locally, a regular GET request is issued, fetching the id:s of the layers needed. This request *does* counts toward the rate limit. Missing layers will then be pulled in parallel.

#### Conclusion

From what I can tell, for most users, no change will be needed and watchtower will continue to work just as before. If the images you monitor change very frequently, you may authenticate to get the rate limit bumped to 200. If you *still* hit the limit, you should probably consider paying for a pro plan.

**Comment by nickperkinslondon** [comment]
I have 20 deployments, each checking every 5 minutes. ( free, private, authenticated on my account, each using different a Access Token )
I only upload a new build maybe twice a week.
So in total, every 6 hours, I should be getting....1440 "checks", and...no more than 20 image downloads.
But I am now starting to hit Rate Limits!
How is it hitting the rate limit if it only does 20 image downloads in one day?  ( and "checks" don't count ? )

**Reply by piksel** [reply]
It's probably easier if you create a new discussion with your logs included. Perhaps the HEAD requests are failing, your auth doesn't work or there are more updates than you think? It's all speculation without the logs of course.

**Comment by simonaronsson** [comment, marked as answer]
HTTP Head requests have now been implemented directly in watchtower as a first comparison to avoid getting rate limited due to  digest checks.

**Reply by yellowbean** [reply]
Does that means, the `watchtower` won't pull actually a image to local if  it knows that the one has been pulled already ?

**Reply by piksel** [reply]
Yes. It will first do a HEAD request to the docker repository, which only returns the hash of the latest image. Only if that hash does not match the one that the container currently use it will be pulled.
